### PR TITLE
Override/redux recompose

### PR DESCRIPTION
**Pull request by EstebanIndiveri**
## Summary

Added useFilterBooks custom Hook filter search books and show matches, if we don't have match show "no Results" message.
Added state searchBook , actionCreactot  and override actionsPrimary in reducer,
destructuring screens by minimal components.
Created keyExtractor utility. 

## Screenshots

[Change!] Show the screenshots of the views you modified.

### Android
<a href="https://ibb.co/cL8YTMz">Screenshot-1622144457</a><br /><a target='_blank' href='https://imgbb.com/'>gg foto</a><br />

<a href="https://ibb.co/LR5Fc3g">Screenshot-1622144474</a><br /><a target='_blank' href='https://imgbb.com/'>gg foto</a><br />

<a href="https://ibb.co/jzPqNq3">Screenshot-1622144485</a><br /><a target='_blank' href='https://imgbb.com/'>gg foto</a><br />

#### Galaxy S (or similar)
<a href="https://ibb.co/zJdcZ8s">Screenshot-1622144500</a><br /><a target='_blank' href='https://imgbb.com/'>gg foto</a><br />

<a href="https://ibb.co/Mfh4888">Screenshot-1622144515</a><br /><a target='_blank' href='https://imgbb.com/'>gg foto</a><br />

<a href="https://ibb.co/Fb7BhJg">Screenshot-1622144524</a><br /><a target='_blank' href='https://imgbb.com/'>gg foto</a><br />

<a href="https://ibb.co/ZW76hyQ">Screenshot-1622144532</a><br /><a target='_blank' href='https://imgbb.com/'>gg foto</a><br />

### iOS
<a href="https://ibb.co/stLmSdh">Simulator-Screen-Shot-i-Phone-SE-2nd-generation-2021-05-27-at-16-37-22</a><br /><a target='_blank' href='https://imgbb.com/'>gg foto</a><br />

<a href="https://ibb.co/xj9Z2rK">Simulator-Screen-Shot-i-Phone-SE-2nd-generation-2021-05-27-at-16-37-47</a><br /><a target='_blank' href='https://imgbb.com/'>gg foto</a><br />

<a href="https://ibb.co/BVt2WVd">Simulator-Screen-Shot-i-Phone-SE-2nd-generation-2021-05-27-at-16-37-56</a><br /><a target='_blank' href='https://imgbb.com/'>gg foto</a><br />

#### iPhone SE (or similar)
<a href="https://ibb.co/6wTpBqj">Simulator-Screen-Shot-i-Phone-8-2021-05-27-at-16-34-28</a><br /><a target='_blank' href='https://imgbb.com/'>gg foto</a><br />

<a href="https://ibb.co/0hMxn1g">Simulator-Screen-Shot-i-Phone-8-2021-05-27-at-16-34-44</a><br /><a target='_blank' href='https://imgbb.com/'>gg foto</a><br />

<a href="https://ibb.co/FYh2Tv9">Simulator-Screen-Shot-i-Phone-8-2021-05-27-at-16-34-54</a><br /><a target='_blank' href='https://imgbb.com/'>gg foto</a><br />

#### iPhone 6/7/8 (or similar)

<a href="https://ibb.co/5LCKnBG">Simulator-Screen-Shot-i-Phone-12-2021-05-27-at-16-35-35</a><br /><a target='_blank' href='https://imgbb.com/'>gg foto</a><br />

<a href="https://ibb.co/TBG4jf4">Simulator-Screen-Shot-i-Phone-12-2021-05-27-at-16-35-46</a><br /><a target='_blank' href='https://imgbb.com/'>gg foto</a><br />

<a href="https://ibb.co/drfWNHL">Simulator-Screen-Shot-i-Phone-12-2021-05-27-at-16-35-53</a><br /><a target='_blank' href='https://imgbb.com/'>gg foto</a><br />

## Trello Card

[Trello CardHookFilter](https://trello.com/c/w6zJ1sqZ/64-n-113-hooks-ii-usememo)

### CODE REVIEW

File: wbooks/src/app/components/CustomIconHeader/index.tsx [hunk 1]
```diff
@@ -1,5 +1,8 @@
 import React from 'react';
 import { Image, ImageSourcePropType } from 'react-native';
+import { useNavigation } from '@react-navigation/native';
+import Routes from '@constants/routes';
+import { TouchableWithoutFeedback } from 'react-native-gesture-handler';
```
Comment (andreszuluaga-wolox) [inline wbooks/src/app/components/CustomIconHeader/index.tsx]: use this from react-native

File: wbooks/src/app/hooks/useFilterBook.tsx [hunk 2]
```diff
@@ -0,0 +1,19 @@
+import { useEffect, useMemo } from 'react';
+import { BookInterface, BookState } from '@interfaces/book';
+import { useDispatch, useSelector } from 'react-redux';
+import actionCreator from '@redux/book/actions';
+
+export default function useFilterBook() {
+  const dispatch = useDispatch();
+  //   const books = useSelector((state: BookState) => state.books);
```
Comment (andreszuluaga-wolox) [inline wbooks/src/app/hooks/useFilterBook.tsx]: remove comments innecesary

File: wbooks/src/app/components/CustomIconHeader/index.tsx [hunk 3]
```diff
@@ -1,5 +1,8 @@
 import React from 'react';
-import { Image, ImageSourcePropType } from 'react-native';
+import { TouchableWithoutFeedback, Image, ImageSourcePropType } from 'react-native';
+import { useNavigation } from '@react-navigation/native';
+import Routes from '@constants/routes';
+import {} from 'react-native-gesture-handler';
```
Comment (mattgle) [inline wbooks/src/app/components/CustomIconHeader/index.tsx]: ?

File: wbooks/src/app/components/CustomIconHeader/index.tsx [hunk 4]
```diff
@@ -8,5 +11,14 @@ interface Props {
 }
 
 export default function CustomIconHeader({ icon }: Props) {
-  return <Image source={icon} style={styles.image} resizeMethod="scale" resizeMode="contain" />;
+  const navigation = useNavigation();
+
+  const hanldePress = () => {
+    navigation.navigate(Routes.SearchBook);
+  };
```
Comment (mattgle) [inline wbooks/src/app/components/CustomIconHeader/index.tsx]:
```suggestion
  const handlePress = () =>     navigation.navigate(Routes.SearchBook);
 
```

File: wbooks/src/app/components/SearchInputBook/index.tsx [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+export default function SearchInputBook() {
+  const dispatch = useDispatch();
+  const { searchBook } = useFilterBook();
+  const handleInput = (search: string) => dispatch(actionCreator.searchBooks(search));
+
+  return (
+    <TextInput
+      style={styles.textInput}
+      placeholder="Input a book"
+      placeholderTextColor="#323232"
```
Comment (mattgle) [inline wbooks/src/app/components/SearchInputBook/index.tsx]: extract into color constants

File: wbooks/src/app/components/SearchInputBook/styles.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import { StyleSheet, Platform } from 'react-native';
+import { white } from '@constants/colors';
+
+const WIDTH_IOS = Platform.OS === 'ios' ? 300 : '100%';
```
Comment (mattgle) [inline wbooks/src/app/components/SearchInputBook/styles.js]: extract and use it
```suggestion
const isIos = Platform.OS === 'ios' ;
```

File: wbooks/src/app/hooks/useFilterBook.tsx [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+export default function useFilterBook() {
+  const dispatch = useDispatch();
+  const { books, searchBook } = useSelector((state: BookState) => state);
+  useEffect(() => {
+    dispatch(actionCreator.getBooksList());
+  }, [dispatch]);
+
+  const filteredBooks: BookInterface[] = useMemo(() => {
+    return books.filter(x => x.title.toLowerCase().includes(searchBook?.toLowerCase()));
```
Comment (mattgle) [inline wbooks/src/app/hooks/useFilterBook.tsx]: change X to book

File: wbooks/src/app/screens/SearchBook/index.tsx [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+import useFilterBook from '@hooks/useFilterBook';
+import keyExtractor from '@utils/keyExtractor';
+
+import styles from './styles';
+
+export default function SearchBook() {
+  const { filteredBooks } = useFilterBook();
+  return (
+    <>
+      {filteredBooks.length === 0 && <NoneResult />}
```
Comment (mattgle) [inline wbooks/src/app/screens/SearchBook/index.tsx]: i think this should work too
```suggestion
      {filteredBooks.length && <NoneResult />}
```

File: wbooks/src/utils/keyExtractor.tsx [hunk 9]
```diff
@@ -0,0 +1,2 @@
+const keyExtractor = (item: { id: number }) => item.id.toString();
```
Comment (mattgle) [inline wbooks/src/utils/keyExtractor.tsx]:
```suggestion
const keyExtractor = (item: { id: number }) =>id.toString();
```

**Review by mattgle — CHANGES_REQUESTED** [review]

File: wbooks/src/app/screens/SearchBook/index.tsx (hunk 8 above)
Comment (EstebanIndiveri) [reply wbooks/src/app/screens/SearchBook/index.tsx]: not worked for my, but i replaced it by const function

File: wbooks/src/app/components/CustomIconHeader/index.tsx [hunk 10]
```diff
@@ -8,5 +10,12 @@ interface Props {
 }
 
 export default function CustomIconHeader({ icon }: Props) {
-  return <Image source={icon} style={styles.image} resizeMethod="scale" resizeMode="contain" />;
+  const navigation = useNavigation();
+
+  const hanldePress = () => navigation.navigate(Routes.SearchBook);
```
Comment (mattgle) [inline wbooks/src/app/components/CustomIconHeader/index.tsx]:
```suggestion
  const handlePress = () => navigation.navigate(Routes.SearchBook);
```

File: wbooks/src/app/components/NoResults/index.tsx [hunk 11]
```diff
@@ -0,0 +1,12 @@
+import React from 'react';
+import { View, Text } from 'react-native';
+
+import styles from './styles';
+
+export default function NoneResult() {
```
Comment (mattgle) [inline wbooks/src/app/components/NoResults/index.tsx]: i think EmptySearch is a better name

**Review by mattgle — CHANGES_REQUESTED** [review]

**Review by mattgle — APPROVED** [review]

**Pull request merged**